mm2src/coins/eth.rs

```diff
@@ -291,6 +331,176 @@ impl Default for EthGasLimit {
     }
 }
 
+#[derive(Default, Deserialize)]
+#[serde(default)]
+pub struct EthGasLimitV2 {
+    pub maker: MakerGasLimitV2,
+    pub taker: TakerGasLimitV2,
+    pub nft_maker: NftMakerGasLimitV2,
+}
+
+#[derive(Deserialize)]
+#[serde(default)]
+pub struct MakerGasLimitV2 {
+    pub eth_payment: u64,
+    pub erc20_payment: u64,
+    pub eth_taker_spend: u64,
+    pub erc20_taker_spend: u64,
+    pub eth_maker_refund_timelock: u64,
+    pub erc20_maker_refund_timelock: u64,
+    pub eth_maker_refund_secret: u64,
+    pub erc20_maker_refund_secret: u64,
+}
+
+#[derive(Deserialize)]
+#[serde(default)]
+pub struct TakerGasLimitV2 {
+    pub eth_payment: u64,
+    pub erc20_payment: u64,
+    pub eth_maker_spend: u64,
+    pub erc20_maker_spend: u64,
+    pub eth_taker_refund_timelock: u64,
+    pub erc20_taker_refund_timelock: u64,
+    pub eth_taker_refund_secret: u64,
+    pub erc20_taker_refund_secret: u64,
+    pub approve_payment: u64,
+}
+
+#[derive(Deserialize)]
+#[serde(default)]
+pub struct NftMakerGasLimitV2 {
+    pub erc721_payment: u64,
+    pub erc1155_payment: u64,
+    pub erc721_taker_spend: u64,
+    pub erc1155_taker_spend: u64,
+    pub erc721_maker_refund_timelock: u64,
+    pub erc1155_maker_refund_timelock: u64,
+    pub erc721_maker_refund_secret: u64,
+    pub erc1155_maker_refund_secret: u64,
+}
+
+impl EthGasLimitV2 {
+    fn gas_limit(
+        &self,
+        coin_type: &EthCoinType,
+        payment_type: EthPaymentType,
+        method: PaymentMethod,
+    ) -> Result<(Address, u64), String> {
```

Comment:
Why return the token address from this function? I think it's better to separate concerns. Also, NFT gas limits can be returned from here and making this not return a result.

Reply:
> Why return the token address from this function?

Did this to reduce code duplication.
There are lot of places where I need to get both `token_address` and `gas_limit`, except one place in `sign_and_broadcast_taker_payment_spend_impl`
![image](https://github.com/user-attachments/assets/0dd1f6ef-705c-4136-82a6-eb4138d1d945)

and both `token_address` and `gas_limit` values require us to do `coin_type` matching. So I decided we can handle these in one function

In addition, we can use `gas_limit` method at the beginning of function which covers `coin_type` matching check, so no need to do it later

> Also, NFT gas limits can be returned from here and making this not return a result.

Nft requires to provide `ContractType`, I decides to separate gas limit methods for fungible and non-fungible tokens.

Reply:
I would rather match again to have better code :)
Also, this function is named `gas_limit` but it does more than that which doesn't make sense. If we refactor this in the future by using generic types, there will be no matching at all and it will be easier to refactor if we separated concerns.

Reply:
> If we refactor this in the future by using generic types, there will be no matching at all and it will be easier to refactor if we separated concerns

I see, it makes sense, will refactor gas_limit tomorrow then.

Reply:
Moved get token address to separate function https://github.com/KomodoPlatform/komodo-defi-framework/pull/2211/commits/1ccfc7748322c4be876c02f59580616b72a4812e